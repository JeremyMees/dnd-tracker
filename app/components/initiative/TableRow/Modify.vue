<script setup lang="ts">
import { INITIATIVE_SHEET } from '~~/constants/provide-keys'

const props = defineProps<{ item: InitiativeSheetRow }>()

const { sheet, update } = validateInject(INITIATIVE_SHEET)

const maxCharacters = computed(() => hasMaxCharacters(sheet.value))

function copyRow(): void {
  if (!sheet.value) return

  const index = getCurrentRowIndex(sheet.value, props.item.id)
  const rows = [...sheet.value.rows]

  if (index === -1) return

  update({
    rows: [
      ...rows,
      { ...props.item, id: crypto.randomUUID() },
    ],
  })
}

function deleteRow(): void {
  if (!sheet.value) return

  const index = getCurrentRowIndex(sheet.value, props.item.id)
  let rows = [...sheet.value.rows]

  if (index === -1) return

  rows = rows.filter(r => r.id !== props.item.id)

  update({ rows })
}
</script>

<template>
  <div class="flex items-center justify-end">
    <NuxtLink
      v-if="item.link"
      v-tippy="$t('components.actions.link')"
      data-test-link
      :aria-label="$t('components.actions.link')"
      class="icon-btn-info"
      target="_blank"
      :to="item.link"
    >
      <Icon
        name="tabler:link"
        :aria-hidden="true"
        class="size-5 min-w-5"
      />
    </NuxtLink>
    <button
      v-tippy="$t('actions.copy')"
      data-test-copy
      :disabled="maxCharacters"
      :aria-label="$t('actions.copy')"
      class="icon-btn-help"
      @click="copyRow"
    >
      <Icon
        name="tabler:copy"
        :aria-hidden="true"
        class="size-5 min-w-5"
      />
    </button>
    <button
      v-tippy="$t('actions.delete')"
      data-test-delete
      :aria-label="$t('actions.delete')"
      class="icon-btn-destructive"
      @click="deleteRow"
    >
      <Icon
        name="tabler:trash"
        :aria-hidden="true"
        class="size-5 min-w-5"
      />
    </button>
  </div>
</template>
