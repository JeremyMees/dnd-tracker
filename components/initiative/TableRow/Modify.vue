<script setup lang="ts">
const props = defineProps<{
  item: InitiativeSheetRow
  sheet: InitiativeSheet | undefined
  update: (payload: Omit<Partial<InitiativeSheet>, NotUpdatable | 'campaign'>) => Promise<void>
}>()

function copyRow(): void {
  if (!props.sheet) return

  const index = getCurrentRowIndex(props.sheet, props.item.id)
  const rows = [...props.sheet.rows]

  if (index === -1) return

  props.update({
    rows: [
      ...rows,
      { ...props.item, id: crypto.randomUUID() },
    ],
  })
}

function deleteRow(): void {
  if (!props.sheet) return

  const index = getCurrentRowIndex(props.sheet, props.item.id)
  let rows = [...props.sheet.rows]

  if (index === -1) return

  rows = rows.filter(r => r.id !== props.item.id)

  props.update({ rows })
}
</script>

<template>
  <div class="flex items-center justify-end">
    <button
      v-tippy="$t('actions.copy')"
      :aria-label="$t('actions.copy')"
      class="icon-btn-info"
      @click="copyRow"
    >
      <Icon
        name="tabler:copy"
        :aria-hidden="true"
        class="size-5 min-w-5"
      />
    </button>
    <button
      v-tippy="$t('actions.delete')"
      :aria-label="$t('actions.delete')"
      class="icon-btn-destructive"
      @click="deleteRow"
    >
      <Icon
        name="tabler:trash"
        :aria-hidden="true"
        class="size-5 min-w-5"
      />
    </button>
  </div>
</template>
