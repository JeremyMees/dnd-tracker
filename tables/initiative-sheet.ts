import { createColumnHelper, type InitialTableState, type Row } from '@tanstack/vue-table'
import { actionsTable, expandButton, iconButton } from './generate-functions'

import {
  InitiativeTableRowName,
  InitiativeTableRowHealth,
  InitiativeTableRowActions,
  InitiativeTableRowConditions,
  InitiativeTableRowNotes,
  InitiativeTableRowInit,
  InitiativeTableRowAc,
  InitiativeTableRowDeathSaves,
  InitiativeTableRowConcentration,
  InitiativeTableRowModify,
} from '#components'

const columnHelper = createColumnHelper<InitiativeSheetRow>()

interface ColumnOptions {
  sheet: ComputedRef<InitiativeSheet | undefined>
  update: (payload: Omit<Partial<InitiativeSheet>, NotUpdatable>) => Promise<void>
  openQuickInitModal: () => void
}

export function generateColumns({ sheet, update, openQuickInitModal }: ColumnOptions) {
  const { t } = useI18n()

  return [
    columnHelper.display({
      enableGlobalFilter: false,
      id: 'index',
      cell: ({ row }) => h('div', { class: 'flex items-center gap-x-2 max-w-20' }, [
        h('span', { class: 'text-muted-foreground ml-1' }, row.index + 1),
        expandButton({
          content: t(`actions.${row.getIsExpanded() ? 'hide' : 'show'}`),
          expanded: row.getIsExpanded(),
          cb: () => row.toggleExpanded(),
        }),
      ]),
    }),
    columnHelper.accessor('name', {
      header: t('components.encounterTable.headers.name'),
      cell: ({ row }) => h(InitiativeTableRowName, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('initiative', {
      header: () => h('button', { class: 'flex items-center gap-x-2' }, [
        t('components.encounterTable.headers.init'),
        iconButton({
          icon: 'tabler:bolt-filled',
          content: t('components.encounterTable.quick'),
          color: 'warning',
          cb: () => openQuickInitModal(),
          disabled: !sheet.value?.rows.length,
          iconColor: 'text-warning',
        }),
      ]),
      cell: ({ row }) => h(InitiativeTableRowInit, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('health', {
      header: t('components.encounterTable.headers.health'),
      cell: ({ row }) => h(InitiativeTableRowHealth, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('ac', {
      header: t('components.encounterTable.headers.ac'),
      cell: ({ row }) => h(InitiativeTableRowAc, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.display({
      id: 'manage',
      header: t('components.encounterTable.headers.manage'),
      cell: ({ row }) => h(InitiativeTableRowActions, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('conditions', {
      header: t('components.encounterTable.headers.conditions'),
      cell: ({ row }) => h(InitiativeTableRowConditions, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('note', {
      header: t('components.encounterTable.headers.note'),
      cell: ({ row }) => h(InitiativeTableRowNotes, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('deathSaves', {
      header: t('components.encounterTable.headers.deathSaves'),
      cell: ({ row }) => h(InitiativeTableRowDeathSaves, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('concentration', {
      header: t('components.encounterTable.headers.concentration'),
      cell: ({ row }) => h(InitiativeTableRowConcentration, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.display({
      enableGlobalFilter: false,
      enableSorting: false,
      id: 'actions',
      cell: ({ row }) => h(InitiativeTableRowModify, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
  ]
}

export function expandedMarkup(row: Row<InitiativeSheetRow>) {
  return actionsTable(row.original)
}

export const initialState: InitialTableState = {}
