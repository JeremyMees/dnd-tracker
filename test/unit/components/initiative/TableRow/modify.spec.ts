import { mountSuspended } from '@nuxt/test-utils/runtime'
import { describe, expect, it, vi, beforeEach } from 'vitest'
import Modify from '~/components/initiative/TableRow/Modify.vue'
import { sheet } from '~~/test/unit/fixtures/initiative-sheet'

interface Props {
  item: InitiativeSheetRow
  sheet: InitiativeSheet | undefined
  update: (payload: Omit<Partial<InitiativeSheet>, NotUpdatable | 'campaign'>) => Promise<void>
}

const mockUpdate = vi.fn()

const props: Props = {
  item: sheet.rows[0]!,
  sheet,
  update: mockUpdate,
}

describe('Initiative table row modify', async () => {
  beforeEach(() => mockUpdate.mockClear())

  it('Should match snapshot', async () => {
    const component = await mountSuspended(Modify, { props })
    expect(component.html()).toMatchSnapshot()
  })

  it('Should display link button when item has link', async () => {
    const component = await mountSuspended(Modify, {
      props: {
        ...props,
        item: { ...props.item, link: 'https://example.com' },
      },
    })

    expect(component.find('[data-test-link]').exists()).toBe(true)
  })

  it('Should not display link button when item has no link', async () => {
    const component = await mountSuspended(Modify, {
      props: {
        ...props,
        item: { ...props.item, link: undefined },
      },
    })

    expect(component.find('[data-test-link]').exists()).toBe(false)
  })

  it('Should call update with new row when copy button is clicked', async () => {
    const component = await mountSuspended(Modify, { props })

    await component.find('[data-test-copy]').trigger('click')

    expect(mockUpdate).toHaveBeenCalledWith({
      rows: expect.arrayContaining([
        expect.objectContaining({
          ...props.item,
          id: expect.any(String),
        }),
      ]),
    })
  })

  it('Should call update with filtered rows when delete button is clicked', async () => {
    const component = await mountSuspended(Modify, { props })

    await component.find('[data-test-delete]').trigger('click')

    expect(mockUpdate).toHaveBeenCalledWith({
      rows: expect.arrayContaining(
        props.sheet!.rows.filter(row => row.id !== props.item.id),
      ),
    })
  })

  it('Should disable copy button when max characters is reached', async () => {
    const maxRows = Array.from({ length: 50 }, () => ({
      ...props.item,
      id: crypto.randomUUID(),
    }))

    const component = await mountSuspended(Modify, {
      props: {
        ...props,
        sheet: {
          ...props.sheet!,
          rows: maxRows,
        },
      },
    })

    expect(component.find('[data-test-copy]').attributes('disabled')).toBeDefined()
  })

  it('Should not call update when sheet is undefined', async () => {
    const component = await mountSuspended(Modify, {
      props: {
        ...props,
        sheet: undefined,
      },
    })

    await component.find('[data-test-copy]').trigger('click')
    await component.find('[data-test-delete]').trigger('click')

    expect(mockUpdate).not.toHaveBeenCalled()
  })
})
