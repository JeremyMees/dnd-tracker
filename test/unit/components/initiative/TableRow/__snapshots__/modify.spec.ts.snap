// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Initiative table row modify > Should match snapshot 1`] = `"<div class="flex items-center justify-end"><a href="https://www.dndbeyond.com/characters/114215312" data-test-link="" aria-label="components.actions.link" class="icon-btn-info" target="_blank"><span class="iconify i-tabler:link size-5 min-w-5" aria-hidden="true"></span></a><button data-test-copy="" aria-label="actions.copy" class="icon-btn-help"><span class="iconify i-tabler:copy size-5 min-w-5" aria-hidden="true"></span></button><button data-test-delete="" aria-label="actions.delete" class="icon-btn-destructive"><span class="iconify i-tabler:trash size-5 min-w-5" aria-hidden="true"></span></button></div>"`;
